packages/protocol/src/utils/StreamPartitionID.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * SPID - Stream Partition ID
+ */
+export default class SPID {
+    /** stream id */
+    public readonly id: string
+    /** stream partition */
+    public readonly partition: number
+    /** toString/fromString separator */
+    protected static readonly SEPARATOR = '|'
+
+    /** string key representing SPID */
+    public readonly key: string
+
+    protected readonly cachedSplit: [string, number] // cached split
+
+    /**
+     * @param id - stream id
+     * @param partition - stream partition
+     */
+    constructor(id: string, partition = 0) {
+        this.id = typeof id === 'string' ? id.toLowerCase() : id
+        this.partition = partition
+        this.validate()
+
+        // static values
+        const separator = (this.constructor as typeof SPID).SEPARATOR
```

Comment:
`SPID` isn't really correct though, it should always read `this.constructor` when referencing static values from the current class, this prevents issues when subclassing.

Reply:
Not that subclassing is likely or desirable here, just you should always reference `this.constructor` as a habit so you never have to think about whether you're going to break subclassing or not.

Reply:
Coming from Java, static things are never overridden by subclasses. I don't know if this is a general principle in other languages as well.

Be it as it may, it feels quite unidiomatic (and hard to remember) to me to have to call static things with `(this.constructor as typeof BaseClass)`, maybe I'd rather then have them as constants outside the class`

E.g.
const SEPARATOR = '|'
...
class SPID {
    ....
}
```

Reply:
>  it feels quite unidiomatic (and hard to remember) 

I guess it feels unidiomatic because people don't subclass much and haven't felt the pain, or they just don't think about how it works? same as most people use `.catch` instead of second argument to `then` in a way that allows double handling or control reentry, or subclass errors without calling `Error.captureStackTrace` or `throw 'some string'`. These all technically "work" but is it good?

Reply:
ahhhh well, I say this but I guess the same problem exists in the calls to `SPID.from` etc. 
OK, I'll just change it to `SPID` here.